### PR TITLE
Added DPMS for GBM/DRMAtomic

### PR DESCRIPTION
**Pull request by lorenz**
## Description
This is an (admittedly pretty crude, this is my first contribution to Kodi) attempt to add support for DPMS (Display Power Management Signaling) to the GBM code, currently only to the atomic modesetting part.

## Motivation and Context
All DPMS-related functionality including sleep timer and ToggleDPMS() for software-based power management doesn't work when using the GBM backend.

## How Has This Been Tested?
Has been tested on a 7th-gen Intel NUC running Linux 4.15 feeding a 4K60Hz HDMI monitor and is working. Has been tested on the Nouveau driver with atomic modesetting disabled and doesn't break anything (but DPMS doesn't work since it is currently not implemented for legacy modesetting).

Should have minimal impact on other code.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by hudokkow** [comment]
Thank you for your contribution and congrats on your first PR to Kodi! Much appreciated.

ping @FernetMenta and @lrusak

File: xbmc/windowing/gbm/DRMAtomic.cpp [hunk 1]
```diff
@@ -140,7 +140,7 @@ void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool video
       return;
     }
 
-    if (!AddCrtcProperty(m_req, m_crtc->crtc->crtc_id, "ACTIVE", 1))
+    if (!AddCrtcProperty(m_req, m_crtc->crtc->crtc_id, "ACTIVE", (int)m_active))
```
Comment (lrusak) [inline xbmc/windowing/gbm/DRMAtomic.cpp]: No c-style casts please

File: xbmc/windowing/gbm/DRMAtomic.h [hunk 2]
```diff
@@ -29,15 +29,18 @@ class CDRMAtomic : public CDRMUtils
   ~CDRMAtomic() { DestroyDrm(); };
   virtual void FlipPage(struct gbm_bo *bo, bool rendered, bool videoLayer) override;
   virtual bool SetVideoMode(RESOLUTION_INFO res, struct gbm_bo *bo) override;
+  virtual bool SetActive(bool active);
   virtual bool InitDrm() override;
   virtual void DestroyDrm() override;
 
   bool AddPlaneProperty(drmModeAtomicReq *req, struct plane *obj, const char *name, int value);
+
```
Comment (lrusak) [inline xbmc/windowing/gbm/DRMAtomic.h]: Uneeded extra line

File: xbmc/windowing/gbm/DRMAtomic.h [hunk 3]
```diff
@@ -29,15 +29,18 @@ class CDRMAtomic : public CDRMUtils
   ~CDRMAtomic() { DestroyDrm(); };
   virtual void FlipPage(struct gbm_bo *bo, bool rendered, bool videoLayer) override;
   virtual bool SetVideoMode(RESOLUTION_INFO res, struct gbm_bo *bo) override;
+  virtual bool SetActive(bool active);
```
Comment (lrusak) [inline xbmc/windowing/gbm/DRMAtomic.h]: Missing override

File: xbmc/windowing/gbm/DRMAtomic.cpp [hunk 4]
```diff
@@ -229,6 +229,7 @@ void CDRMAtomic::FlipPage(struct gbm_bo *bo, bool rendered, bool videoLayer)
 
 bool CDRMAtomic::InitDrm()
 {
+  m_active = true;
```
Comment (lrusak) [inline xbmc/windowing/gbm/DRMAtomic.cpp]: Can you please set this in the header file instead?

**Review by lrusak — CHANGES_REQUESTED** [review]

**Comment by lrusak** [comment]
Generally it looks pretty good! I'll have to give it a test later. I'll also look at what is required for legacy drm. Thanks for the first contribution!

File: .gitignore [hunk 5]
```diff
@@ -165,6 +165,7 @@ cmake_install.cmake
 /lib/cpluff/loader/cpluff-loader
 /lib/cpluff/m4/
 /lib/cpluff/stamp-h1
+/lib/cpluff/tmp*
```
Comment (Rechi) [inline .gitignore]: unrelated to this PR, if this is needed move it at least into it's own commit (better would be a separate PR)

File: .gitignore [hunk 6]
```diff
@@ -352,4 +353,4 @@ exclude_dll.txt
 
 # VS git tools doesn't work for true out of tree builds and it's
 # useful to be able to work on multiple configs
-/build*/
+/build*
```
Comment (Rechi) [inline .gitignore]: unrelated, same as above

**Comment by lrusak** [comment]
After looking at this again I think the approach is wrong. I had to google a bit but we should be setting the DPMS property in the connector properties. This will then support standby, suspend, off, and on.

I have created a commit that does these changes. You mostly had the right idea though. I have tested this on legacy DRM (radeon) but will still need to test it on atomic drm.

If you agree please pick and squash all commits, https://github.com/lrusak/xbmc/commit/25b5ee005bd8ce146ea88d25e9a5cea14c74594f

I think this still can be improved by actually detecting the DPMS properties that are supported but this is a good start.

**Comment by lorenz** [comment]
On atomic DPMS is deprecated and being translated to the ACTIVE property on the CRTC as far as I read the Linux code and docs

**Comment by lrusak** [comment]
Yep I see that now.

```
Legacy entry point to set the per-connector DPMS state. Legacy DPMS is exposed as a standard property on the connector, but diverted to this callback in the drm core. Note that atomic drivers don’t implement the 4 level DPMS support on the connector any more, but instead only have an on/off “ACTIVE” property on the CRTC object.
```

So what you have done is ok but I'd still like to squash some stuff up and clean up other things.

**Comment by lrusak** [comment]
@lorenz can you please pick and squash https://github.com/lrusak/xbmc/commit/f542f060522c9755a967cf548c272b754d9c5984

and then just pick https://github.com/lrusak/xbmc/commit/7eed15c595cd6afb4967532208aa87fbb2a15ed5

I've tested both atomic and legacy drm and this is working fine

**Comment by lorenz** [comment]
@lrusak I squashed and picked your commit, should be all good now

**Review by lrusak — APPROVED** [review]

**Comment by lrusak** [comment]
thanks for your contribution!

**Comment by Rechi** [comment]
@lrusak Jenkins hasn't finished building this PR and now GBM is compile broken.

File: xbmc/windowing/gbm/DRMAtomic.cpp [hunk 7]
```diff
@@ -140,7 +140,7 @@ void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool video
       return;
     }
 
-    if (!AddCrtcProperty(m_req, m_crtc->crtc->crtc_id, "ACTIVE", 1))
+    if (!AddCrtcProperty(m_req, m_crtc->crtc->crtc_id, "ACTIVE", m_active ? 1 : 0)
```
Comment (Rechi) [inline xbmc/windowing/gbm/DRMAtomic.cpp]: `? 1 : 0))`

Comment (lorenz) [reply xbmc/windowing/gbm/DRMAtomic.cpp]: Sorry, was just a one-liner, I forgot to locally build it first. But I see you already fixed it in master, thanks!

Comment (Rechi) [reply xbmc/windowing/gbm/DRMAtomic.cpp]: Not your fault, we have CI for that, but should be also wait for it to finish.
